File: Kuznetsov_Ilja/1376.py
from sys import setrecursionlimit
from typing import List

setrecursionlimit(10**5)

class Solution:
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        # construct tree
        children: List[List[int]] = [[] for _ in range(n)]
        for employee_idx, manager_idx in enumerate(manager):
            if manager_idx != -1:
                children[manager_idx].append(employee_idx)
        
        # DFS
        return Solution.DFS(children, informTime, headID)

    @staticmethod
    def DFS(children: List[List[int]], informTime: List[int], current_node_idx: int):
        if children[current_node_idx]:
            max_child_time = max(Solution.DFS(children, informTime, child_idx)
                                for child_idx in children[current_node_idx])
            return informTime[current_node_idx] + max_child_time
        return 0

print(Solution().numOfMinutes(n = 6, headID = 2, manager = [2,2,-1,2,2,2], informTime = [0,0,1,0,0,0]))

File: Leetcode/1376.go
func DFS(tree [][]int, informTime []int, cur_id int) int {
	totalTime := 0
	for i := 0; i < len(tree[cur_id]); i++ {
        next_id:=tree[cur_id][i]
        newTime:=informTime[cur_id]
        if len(tree[next_id]) != 0 {
            newTime = DFS(tree, informTime, next_id)+informTime[cur_id]
        }
		totalTime = int(math.Max(float64(totalTime), float64(newTime)))
	}
	return totalTime
}

func numOfMinutes(n int, headID int, manager []int, informTime []int) int {
    //созданим двумерный массив, где для i-ого человека будет сопоставлен массив с номерами его подчинённых
	tree := make([][]int, n)
	for i := 0; i < n; i++ {
		if i != headID {
			tree[manager[i]] = append(tree[manager[i]], i)
		}
	}
    //пройдёмся по структуре дерева с помощью DFS метода
	totalTime := DFS(tree, informTime, headID)
	return totalTime
}